[feature] Implements CachePolicy: kWTinyLFU, kTinyLFU, kSLRU

raslboyy opened the pull request:
Relates: [ISSUE-205](https://github.com/userver-framework/userver/issues/205)

LRU is a basic and simple caching policy. But in some cases, this is far from the optimal algorithm. We want to present our implementation of modern caching policies and compare them with LRU. 

Our incremental cache improvements:
- SLRU
Implemented with two basic LRU from userver.  

- TinyLfu
Consists of *proxy* and *main* parts. *Proxy* - frequency counter approximation. *Main* - any caching algorithm that implements `LruBase` interface. Evicted items from *main* compared in frequency with new candidate for put.

- WTinyLfu
Consists of *window* and *main* parts. *Window* - basic LRU from userver. *Main* - TinyLfu. All evicted items from *window* are putted to *main*.

Frequency sketches:
- Bloom
- Bloom with Doorkeeper

Implemented as template specialization of `LruBase` with addition new non-type template parameter . All caches kept the `LruBase` interface unchanged. 

More details in [paper](https://arxiv.org/pdf/1512.00727.pdf)

----------

![image](https://user-images.githubusercontent.com/33222839/205645140-e8c2a09d-82c4-4ff3-a9ff-2f7cd04665b2.png)

stewkk committed:
add simple lfu and benchmark

Commit:
add ci-cd

LeonidFaktorovich committed:
added intrusive map

LeonidFaktorovich committed:
added pragma once

LeonidFaktorovich committed:
deleted m_map_

LeonidFaktorovich committed:
use clear in destuctor

raslboyy committed:
feat: add cache policy parametrization to LRUCacheComponent

raslboyy committed:
refactor: add last line, change include

raslboyy committed:
fix: compile error)

raslboyy committed:
add: policy enum

raslboyy committed:
add: draft tiny_lfu and implemented BloomFilter for frequency sketching

raslboyy committed:
add: base logic test for bloom

raslboyy committed:
refactor: bloom_test

raslboyy committed:
add: more tests for bloom

raslboyy committed:
refactor: bloom

raslboyy committed:
add: draft tiny_lfu

raslboyy committed:
refactor: bloom

raslboyy committed:
refactor: bloom

LeonidFaktorovich committed:
add clear method for bloom and added implementation for kTinyLFU cache policy

LeonidFaktorovich committed:
added closing comment for namespace cache

raslboyy committed:
refactor: remove new

raslboyy committed:
refactor:

raslboyy committed:
Merge remote-tracking branch 'github/feature-core-tinylfu' into feature-core-tinylfu

# Conflicts:
#	shared/include/userver/cache/impl/tiny_lfu.hpp

raslboyy committed:
add: slru v1 and some basic tests

raslboyy committed:
fix: incorrect get least item

LeonidFaktorovich committed:
added move method and move with set value method in lru base

raslboyy committed:
add: basic tests to TinyLFU; fix: push and emplace to TinyLFU; refactor: simplified capacity

LeonidFaktorovich committed:
fix: template friend in lru base

raslboyy committed:
refactor: clang-tidy

raslboyy committed:
fix: includes and GetCacheStatisticsAsJson

raslboyy committed:
add: new test case -- fix: SLRU bug if item already in protected section

raslboyy committed:
fix: bug empty protected LRU

raslboyy committed:
fix: bug SetMaxSize

raslboyy committed:
add: parametrized NWayLRU tests

raslboyy committed:
refactor: NWayLRU test

raslboyy committed:
refactor: NWayLRU test

raslboyy committed:
refactor: NWayLRU test

raslboyy committed:
refactor: NWayLRU test

raslboyy committed:
add: parametrized ExpirableLRU tests

raslboyy committed:
feat: parametrize hash in frequency sketch

raslboyy committed:
feat: separate hash functions

raslboyy committed:
add: doorkeeper and basic tests

raslboyy committed:
add: DoorkeeperBloom implementation + some tests

raslboyy committed:
refactor: FrequencySketchF.Full only for bloom

raslboyy committed:
feat: parametrize inner cache policy in tinyLFU

raslboyy committed:
refactor: unbind tinylfu impl from lru implementation

raslboyy committed:
feat: parametrize TinyLFU tests

LeonidFaktorovich committed:
add: using hash and equal in tests

LeonidFaktorovich committed:
add: using hash and equal in tests

LeonidFaktorovich committed:
add: using hash and equal in core tests

raslboyy committed:
add: window tiny lfu add some basic tests

raslboyy committed:
add: window tiny lfu add some basic tests

raslboyy committed:
refactor: tests

raslboyy committed:
fix: includes

LeonidFaktorovich committed:
change put method in slru

LeonidFaktorovich committed:
change put method in slru

LeonidFaktorovich committed:
add: branch feature-core-slru to ci

LeonidFaktorovich committed:
add: branch feature-core-slru to ci

LeonidFaktorovich committed:
fix: move element in put from protected to probation

LeonidFaktorovich committed:
fix: move element in put from protected to probation

raslboyy committed:
fix: slru get least

stewkk committed:
add slru and tinylfu benches

raslboyy committed:
feat: enable doorkeeper

raslboyy committed:
Revert "add: window tiny lfu add some basic tests"

This reverts commit da6e943d22c021d8802cd9b24704406f723bb7cb.

LeonidFaktorovich committed:
change: get method in slru

LeonidFaktorovich committed:
Merge branch 'feature-core-slru' of github.com:chistopat/userver into feature-core-slru

# Conflicts:
#	shared/include/userver/cache/impl/slru.hpp

stewkk committed:
add cpu benchmarks on real data samples

chistopat committed:
Merge branch 'userver-framework:develop' into develop

Commit:
Merge branch 'develop' into feature-core-wtinylfu

stewkk committed:
Refactor slru and tinylfu benchmarks

stewkk committed:
Merge branch 'feature-core-slru' into feature-core-benchmarks

stewkk committed:
Merge remote-tracking branch 'origin/feature-core-wtinylfu' into feature-core-benchmarks

stewkk committed:
Add WTinyLfu bench

stewkk committed:
Add bench runner

raslboyy committed:
feat: edit some default params and round capacity to Next power of two

raslboyy committed:
feat: edit some default params and round capacity to Next power of two

raslboyy committed:
feat: edit some default params and round capacity to Next power of two

raslboyy committed:
refactor: frequency logic

raslboyy committed:
refactor: frequency logic

raslboyy committed:
refactor: frequency logic

raslboyy committed:
fix: default probation_part

raslboyy committed:
fix: default probation_part

raslboyy committed:
fix: default probation_part

raslboyy committed:
refactor: TinyLfu

raslboyy committed:
refactor: TinyLfu

raslboyy committed:
refactor: TinyLfu

raslboyy committed:
fix: incorrect window tiny lfu, add RecordAccess for window

raslboyy committed:
fix: incorrect window tiny lfu, add RecordAccess for window

raslboyy committed:
fix: incorrect window tiny lfu, add RecordAccess for window

raslboyy committed:
refactor

raslboyy committed:
refactor

raslboyy committed:
refactor

raslboyy committed:
add: new implementation sketch counters

raslboyy committed:
add: new implementation sketch counters

raslboyy committed:
add: new implementation sketch counters

raslboyy committed:
add: doorkeeper

raslboyy committed:
add: doorkeeper

raslboyy committed:
add: doorkeeper

raslboyy committed:
add: aged -- sketch with auto reset

raslboyy committed:
add: aged -- sketch with auto reset

raslboyy committed:
add: aged -- sketch with auto reset

raslboyy committed:
add: sketch tools impl

raslboyy committed:
add: sketch tools impl

raslboyy committed:
add: sketch tools impl

raslboyy committed:
add: sketch tests

raslboyy committed:
add: sketch tests

raslboyy committed:
add: sketch tests

raslboyy committed:
add: new aged sketch test

raslboyy committed:
add: new aged sketch test

raslboyy committed:
add: new aged sketch test

raslboyy committed:
refactor: adapt tiny lfu to new sketch impl

raslboyy committed:
refactor: adapt tiny lfu to new sketch impl

raslboyy committed:
refactor: adapt tiny lfu to new sketch impl

raslboyy committed:
delete old sketch

raslboyy committed:
delete old sketch

raslboyy committed:
delete old sketch

raslboyy committed:
refactor

raslboyy committed:
refactor

raslboyy committed:
refactor

raslboyy committed:
refactor: split TinyLFU into internal and external partrs, add: mocked sketch, fix: TinyLFU to mocked sketch

raslboyy committed:
refactor: split TinyLFU into internal and external partrs, add: mocked sketch, fix: TinyLFU to mocked sketch

raslboyy committed:
refactor: split TinyLFU into internal and external partrs, add: mocked sketch, fix: TinyLFU to mocked sketch

LeonidFaktorovich committed:
add: branch to ci

LeonidFaktorovich committed:
add: branch to ci

LeonidFaktorovich committed:
add: method move if has and move if has with set value in slru

LeonidFaktorovich committed:
add: method move if has and move if has with set value in slru

LeonidFaktorovich committed:
Merge branch 'feat-new-counters-implementation' of github.com:chistopat/userver into feat-new-counters-implementation

# Conflicts:
#	.github/workflows/ci.yml
#	shared/include/userver/cache/impl/slru.hpp

raslboyy committed:
edit end of line for stewkk

raslboyy committed:
fix: include bug stewkk

raslboyy committed:
edit end of line for stewkk

LeonidFaktorovich committed:
Merge branch 'feat-new-counters-implementation' of github.com:chistopat/userver into feat-new-counters-implementation

# Conflicts:
#	shared/include/userver/cache/impl/base.hpp
#	shared/include/userver/cache/impl/hash.hpp
#	shared/include/userver/cache/impl/slru.hpp
#	shared/include/userver/cache/impl/tiny_lfu.hpp
#	shared/include/userver/cache/impl/window_tiny_lfu.hpp

LeonidFaktorovich committed:
add: caffeine bloom policy

stewkk committed:
fix: remove absolute path to test data

stewkk committed:
add: hit rate metric

raslboyy committed:
refactor: TinyLfu

raslboyy committed:
refactor: TinyLfu

stewkk committed:
add: wtinylfu benchmarks

raslboyy committed:
fix: compile bug

raslboyy committed:
fix: compile bug

ben-manes commented:
fyi, I recently [optimized](https://github.com/ben-manes/caffeine/blob/19163d93ac6572f0d19224842ef72db4c4fe557b/caffeine/src/main/java/com/github/benmanes/caffeine/cache/FrequencySketch.java#L42-L49) Caffeine's frequency sketch which yours is based on. It now takes into account the cpu cache line to reduce 4 random reads to 1 or 2 (depending on alignment). It was already fast enough, so just a fun change to avoid being unnecessarily wasteful.

stewkk committed:
add: zipfian dataset

stewkk committed:
add: plots for cpu benchmarks

segoon committed:
bug server: fix args parsing speed

Anton3 committed:
fix dump: enforce specifying dump.min-interval for LRU caches

dengaleev committed:
fix formats::yaml: exception message typo

Tests: протестировано CI

Pull Request resolved: #201

lexlz committed:
feat tests: add mongo chaos tests

segoon committed:
feat passenger-authorizer: implement stream API

Anton3 committed:
fix utils: don't try to lower-case non-letter characters

Anton3 committed:
refactor cache: move dump integration into NWayLRU

LeonidFaktorovich committed:
add: move if has and move if has with set value in tiny lfu

segoon committed:
bug server: fix race in stream error handling

segoon committed:
feat samples, functional_tests: generate fallback files

Anton3 committed:
fix testsuite: allow using dynamic_config fixture out-of-the-box

Anton3 committed:
refactor dynamic_config: remove left-over uses of taxi_config naming

Commit:
feat uservices-linters: add linters test for common code

segoon committed:
bug core: add lost config

itrofimow committed:
fix build: pin postgres version in GitHub CI

Pull Request resolved: #208

itrofimow committed:
feat encoding: use SSE3 for ToHex

Pull Request resolved: #207

segoon committed:
feat redis: add master migration tests

itrofimow committed:
fix rabbitmq: fix subtle race in samples

Tests: протестировано CI

Pull Request resolved: #210

Anton3 committed:
refactor userver: fix C++20 incompatibilities and warnings

- fmt 8+ requires `fmt::runtime` around format strings if `consteval` is available
- `FMT_STRING` is deprecated in fmt 8+ if `consteval` is available
- Make `formatter<TimeOfDay>::parse` `constexpr` to avoid requiring the usage of `fmt::runtime`
- Fix the following warning on `storages::postgres::detail::ConstDataIterator::operator!=`:
   - `warning: ISO C++20 considers use of overloaded operator '!=' to be ambiguous`

segoon committed:
bug redis: don't expose arcadia/ to github

Anton3 committed:
feat postgres: add ProjectedUnorderedSet to avoid key duplication in pg caches

segoon committed:
feat congestion-control: better activated metric value

Anton3 committed:
fix tests: fix TaskInheritedVariable.VariablesAfterParentTaskDeath

Commit:
feat bank-authproxy: add set-cookie processing to response

Anton3 committed:
refactor mongo: remove DropResult

apolukhin committed:
feat docs: add changelogs for November

Tests: протестировано CI

Anton3 committed:
refactor mongo: migrate to new dynamic_config API

apolukhin committed:
feat docs: switch to the winter logo

Tests: на прод не влияет

Anton3 committed:
feat mongo: add MONGO_DEADLINE_PROPAGATION_ENABLED config

Anton3 committed:
fix mongo: regenerate test dynamic configs

Anton3 committed:
fix mongo: make gtests reentrable

`--gtest_repeat=N` should work with this patch

Anton3 committed:
fix postgres: don't crash pgtest if DB is down

Commit:
feat docs: switch to the winter logo open source

Releates:

Anton3 committed:
refactor redis: make (some) tests reentrable

- Make client_redistest reentrable by clearing the database in-between tests
- Remove code duplication in sentinel_redistest
- Refactor redis/impl/request.hpp
- Make `Command` a proper struct
- Move `Command` to `src`

Anton3 committed:
refactor userver: use TaskCancellationToken instead of internal API

itrofimow committed:
feat core: add a config option to disable tracing http headers

These headers aren't always needed, and they take ~120 bytes per response, which is significant when RPS is in hundreds of thousands, such as in TechEmpower benchmarks

Tests: протестировано CI

Pull Request resolved: #211

Anton3 committed:
fix redis: fix spurious crashes in impl/request.cpp

Anton3 committed:
bug core: don't collect too many stacktraces in production

segoon committed:
feat market-perks: implement YANDEX_PLUS perk

Commit:
feat core: secdist provider

Tests: CI

segoon committed:
feat pass-auth: fix stream API race

apolukhin committed:
feat build: remove unused target aliases

Tests: протестировано CI

vitek committed:
feat testsuite: log colorization support

Tests: протестировано CI

Pull Request resolved: #212

apolukhin committed:
feat build: remove unused target aliases

Tests: протестировано CI

segoon committed:
feat core: implement LazyValue

segoon committed:
bug tests: fix github CI

lexlz committed:
feat redis: command metrics collection

Commit:
fix authproxy: revert multiple cookies

Prev ticket:

Revert "feat bank-authproxy: add set-cookie processing to response"
This reverts commit 37bac013e3639ae95a3285f8fefb4eeeb5512278, reversing
changes made to 7dd76b9bb7f14170c53d1db0616363d03dc98217.

Anton3 committed:
refactor scripts: remove some legacy tier1 scripts

Also further clean up Makefile

lexlz committed:
Revert "feat redis: command metrics collection"


This reverts commit b4aa64acc61d355efc34fda1d99e24a2c47735d4, reversing
changes made to 6e49d30bc386454e60a864c8ea9c4b67113a7f01.

Commit:
feature integraiton tests: fix userver plugins import

fix tests_plugins

apolukhin committed:
feat docs: add auth sample

Tests: протестировано CI

Anton3 committed:
refactor all: remove legacy /taxi/uservices/submodules/testsuite

apolukhin committed:
feat postgres: clarify and update docs

Tests: на прод не влияет

Anton3 committed:
feat testsuite: deduplicate functional tests, enable dynamic config mocks

- Add mocks and fixtures required to test dynamic config in userver testsuite by default
- Test dynamic_config fixture in samples/testsuite-support
- Fix core/functional_tests/basic_chaos
- Fix redis/functional_tests/basic_chaos
- Simplify running functional tests internally

Commit:
refactor codegen: remove debian build dependencies

Commit:
task driver-work-rules: search work rules by name case insensitive

Tests: Протестировано в testsuite и в тестинге

apolukhin committed:
feat deptrans-driver-status: reset metrics properly

Tests: протестировано CI

Commit:
feat clickhouse: add UInt16 support

Commit:
fix grpc: make grpc client methods const

segoon committed:
feat server: add test for cookie parsing speed

apolukhin committed:
feat docs: add navigation README.md files into important directories

Tests: протестировано CI, на прод не влияет

apolukhin committed:
feat testsuite: remove reset_metrics variable from tests_control

Tests: протестировано CI, на прод не влияет

apolukhin committed:
feat postgres: add postgres metrics tests

Tests: протестировано CI, на прод не влияет

LeonidFaktorovich committed:
Merge branch 'feature/tiny-lfu' into feat-new-counters-implementation

chistopat committed:
Merge branch 'userver-framework:develop' into feat-new-counters-implementation

LeonidFaktorovich committed:
fix: return policy in NWayLru and LruCacheComponent
